pkg/server/server_test.go

```diff
@@ -0,0 +1,119 @@
+package server
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	apitypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/metrics/pkg/apis/metrics"
+
+	"sigs.k8s.io/metrics-server/pkg/api"
+	"sigs.k8s.io/metrics-server/pkg/scraper"
+	"sigs.k8s.io/metrics-server/pkg/storage"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+func TestServer(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Server Suite")
+}
+
+var _ = Describe("Server", func() {
+	var (
+		resolution time.Duration
+		server     *server
+		scraper    *scraperMock
+		store      *storageMock
+	)
+
+	BeforeEach(func() {
+		resolution = 60 * time.Second
+		scraper = &scraperMock{
+			result: &storage.MetricsBatch{
+				Nodes: []storage.NodeMetricsPoint{
+					{
+						Name: "node1",
+						MetricsPoint: storage.MetricsPoint{
+							Timestamp:   time.Now(),
+							CpuUsage:    resource.Quantity{},
+							MemoryUsage: resource.Quantity{},
+						},
+					},
+				},
+			},
+		}
+		store = &storageMock{}
+		server = NewServer(nil, nil, nil, store, scraper, resolution)
+	})
+
+	It("liveness should pass before first scrape tick finishes", func() {
+		Expect(server.CheckLiveness(nil)).To(Succeed())
+	})
+	It("liveness should pass if scrape fails", func() {
+		scraper.err = fmt.Errorf("failed to scrape")
+		server.tick(context.Background(), time.Now())
+		Expect(server.CheckLiveness(nil)).To(Succeed())
+	})
+	It("liveness should pass if scrape succeeds", func() {
+		server.tick(context.Background(), time.Now().Add(-resolution))
```

Comment:
As these tests are heavily time dependant, what do you think of fixing the current time value to avoid any flakes?

Reply:
Here I made a shortcut to use internal function `tick` instead of using `RunUntil` which would require mocking whole `time` (including `time.Ticker` which is not trivial). 

I think that using `time.Now` here is ok because of chances of flake are very small. Scraping period (or `metric resolution`) is on magnitude of tens of seconds (60 seconds here) which give huge margine of error to just run couple of lines of code.